internal/exec/stages/disks/luks.go

```diff
@@ -294,36 +300,58 @@ func (s *stage) createLuks(config types.Config) error {
 			// pass the device to clevis. We have to loop each device as
 			// the devices could be on different NICs that haven't come
 			// up yet.
+
+			// List of tang servers proven to be reachable.
+			tangServersVerifiedToBeOnline := []types.Tang{}
 			for _, tang := range luks.Clevis.Tang {
 				u, err := url.Parse(tang.URL)
 				if err != nil {
 					return fmt.Errorf("parsing tang URL: %v", err)
 				}
 				u.Path = path.Join(u.Path, "adv")
-				_, err = s.Fetcher.FetchToBuffer(*u, resource.FetchOptions{})
-				if err != nil {
-					return fmt.Errorf("fetching tang advertisement: %v", err)
+				data, err := s.Fetcher.FetchToBuffer(*u, resource.FetchOptions{})
+				if util.NilOrEmpty(tang.Advertisement) {
+					if err != nil {
+						return fmt.Errorf("fetching tang advertisement: %v", err)
+					}
+				} else {
+					// Check if fetch errored out. If it did, we will have to assume the advertisement is correct.
+					if err != nil {
+						continue
+					}
+					fetchedAdv := string(data)
+					// Since the server is reachable, lets double check that the config advertisement matches the online server's advertisement.
+					if fetchedAdv != *tang.Advertisement {
+						return fmt.Errorf("tang advertisement does not match: %v", err)
+					}
 				}
+				// If we made it this far, the tang server information is valid, and reachable.
+				tangServersVerifiedToBeOnline = append(tangServersVerifiedToBeOnline, tang)
 			}
 
+			// lets bind our device
 			if _, err := s.Logger.LogCmd(
 				exec.Command(distro.ClevisCmd(), "luks", "bind", "-f", "-k", keyFilePath, "-d", devAlias, pin, config), "Clevis bind",
 			); err != nil {
 				return fmt.Errorf("binding clevis device: %v", err)
 			}
 
-			// close & re-open Clevis devices to make sure that we can unlock them
-			if _, err := s.Logger.LogCmd(
-				exec.Command(distro.CryptsetupCmd(), "luksClose", luks.Name),
-				"closing clevis luks device %v", luks.Name,
-			); err != nil {
-				return fmt.Errorf("closing luks device: %v", err)
-			}
-			if _, err := s.Logger.LogCmd(
-				exec.Command(distro.ClevisCmd(), "luks", "unlock", "-d", devAlias, "-n", luks.Name),
-				"reopening clevis luks device %s", luks.Name,
-			); err != nil {
-				return fmt.Errorf("reopening luks device %s: %v", luks.Name, err)
+			// Check if we can safely close and re-open the device
+			if len(tangServersVerifiedToBeOnline) > *luks.Clevis.Threshold || luks.Clevis.Tang != nil && *luks.Clevis.Tpm2 {
```

Comment:
I tried to account for TPM here, I am not sure if this is the best way; as the conditional gets a bit gnarly